### DISCUSSION TITLE
Access ClientContext values in Xamarin.Forms after call from native Android

### DISCUSSION THREAD

**Discussion by tkyle** [MarimerLLC/csla#2504] [Questions]
Let's say I have a value that exists in the ClientContext:

`Csla.ApplicationContext.ClientContext["TestKey"] = "Client Context Value";`

When I access that value in shared code from a call initiated in shared code, that value exists. If I try to access that value from shared code in a call initiated in native code, it does not exist. 

As an example, let's say there's a Firebase specific override that's initiated by the Firebase SDK. If I want to then make a call from that override (native), into shared code, using a static method in shared code, the ClientContext is missing the value.

Is this expected? Am I doing something wrong?

I've attached a sample of what I'm talking about in case it's not expected behavior and what I'm saying is hard to follow.

[ClientContextTest.Mobile.zip](https://github.com/MarimerLLC/csla/files/7239522/ClientContextTest.Mobile.zip)

**Comment by rockfordlhotka** [comment]
It is important to understand how `ClientContext` works.

It is maintained (by default) in an `AsyncLocal` variable, which means that it is available on the parent thread, and should flow through to child threads if/when they are created using things like `Task.Run`.

The data portal flows `ClientContext` through to a remote server by serializing the context on the client, and deserializing it on the server.

I suspect that when the Android native code calls back into your app, that the code in your app isn't running on a child thread of the original parent, and so has no access to `ClientContext`.

Because an Android app is basically just like any other smart client app, you _might_ be able to create your own context manager that stores `ClientContext` in a `static`, making it global to the app's process. In older versions of CSLA (like 4) I think this was the approach used in WPF - so in Csla.Xaml. I say this, because you can find an example of the context manager implementation I'm talking about.

I still don't know for sure if this will work, but it is probably worth a try.

**Reply by tkyle** [reply]
That makes sense. Thanks for the reply.

I will look into this and see what I come up with.
